File: WindowDemo/WindowDemo/CustomBubbleView.m
//
//  CustomBubbleView.m
//  WindowDemo
//
//  Created by Theo on 2024/8/22.
//
#import "CustomBubbleView.h"
#import "Masonry/Masonry.h"
#import "BubbleViewModel.h"
#import "UIColor+LAUIKit.h"

#define DESIGN_WIDTH 750.0
#define LAZSAO750(value) (value * (UIScreen.mainScreen.bounds.size.width / DESIGN_WIDTH))
@interface CustomBubbleView ()
@property (nonatomic, strong) UILabel *label;
@property (nonatomic, strong) UIImageView *leftImageView;
@property (nonatomic, strong) CAShapeLayer *shapeLayer;
@property (nonatomic, strong) CAGradientLayer *gradientContentLayer;
@property (nonatomic, strong) CAGradientLayer *gradientBorderLayer;
@property (nonatomic, strong) CAShapeLayer *borderLayer;

@end

@implementation CustomBubbleView
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupView];
    }
    return self;
}

- (void)setupView{

    
    self.label = [UILabel new];
    [self addSubview:self.label];
    
    self.leftImageView = [[UIImageView alloc] init];
    [self addSubview:self.leftImageView];
    
    self.shapeLayer = [CAShapeLayer layer];
    self.shapeLayer.fillColor = [UIColor whiteColor].CGColor; // 填充颜色
    self.shapeLayer.strokeColor = [UIColor whiteColor].CGColor;
    
    self.gradientContentLayer = [CAGradientLayer layer];
    self.gradientContentLayer.frame = self.bounds;
    self.gradientContentLayer.mask = self.shapeLayer;
    [self.layer insertSublayer:self.gradientContentLayer atIndex:0];
    
    self.borderLayer = [CAShapeLayer layer];
    [self.layer addSublayer:self.borderLayer];
    
    
    [self setupLayout];
}

- (void)bindViewModel:(BubbleViewModel *)viewModel {
    _viewModel = viewModel;
    
    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] init];
    //self.label.text = _viewModel.model.title;
    for(NSDictionary* dict in self.viewModel.model.titleExtend){
        NSString *text = dict[@"text"];
        UIColor *textColor = [UIColor laz_colorWithHex:dict[@"color"]];
        UIFont *font = [UIFont systemFontOfSize:LAZSAO750(24.0)];
        if ([dict[@"bold"] isEqualToString:@"true"]) {
            font = [UIFont boldSystemFontOfSize:LAZSAO750(22.0)];
        }
        NSDictionary *attributes = @{
            NSFontAttributeName: font,
            NSForegroundColorAttributeName: textColor
        };
        NSAttributedString *attributedText = [[NSAttributedString alloc] initWithString:text attributes:attributes];
        [attributedString appendAttributedString:attributedText];
    }
    self.label.numberOfLines = 2;
    self.label.attributedText = attributedString;
    
    self.borderLayer.strokeColor = self.viewModel.model.borderLineColor.CGColor;
    self.borderLayer.lineWidth = self.viewModel.model.borderLineSize;
    self.leftImageView.image = [UIImage imageNamed:@"Image-1614"];
    
    [self setupLayer];
}

- (void)layoutSubviews{
    [super layoutSubviews];
    self.shapeLayer.path = [self createPath].CGPath;
    self.gradientContentLayer.frame = self.bounds;
    self.borderLayer.path = self.shapeLayer.path;
}
-(void)setupLayer{
    self.gradientContentLayer.colors = self.viewModel.model.contentColors;
    // 设置渐变的起始和结束点
    self.gradientContentLayer.startPoint = CGPointMake(0.0, 1.0);
    self.gradientContentLayer.endPoint = CGPointMake(1.0, 0.0);
    // 将ShapeLayer作为gradientContentLayer的蒙版
    self.gradientContentLayer.mask = self.shapeLayer;
    CGFloat locationSize = 1.0 / self.viewModel.model.contentColors.count;
    self.gradientContentLayer.locations = @[@0.0, @0.4 ,@0.7, @1.0];
    
    self.borderLayer.path = self.shapeLayer.path;  // 使用相同的路径
    self.borderLayer.fillColor = [UIColor clearColor].CGColor; // 边界描边不需要填充
}



//modify 比例
-(void)setupLayout {
    [self.leftImageView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.height.equalTo(self.mas_height).offset(-LAZSAO750(12.0 + 10.0) - self.viewModel.model.triangleHeight);
        make.width.equalTo(self.leftImageView.mas_height);
        make.left.equalTo(self).offset(LAZSAO750(12.0));
        make.top.equalTo(self).offset(LAZSAO750(12));
        //make.centerY.equalTo(self);
    }];
    
    [self.label mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.right.equalTo(self).offset(LAZSAO750(-36.0));
        make.left.equalTo(self.leftImageView.mas_right).offset(LAZSAO750(12.0));
        make.top.equalTo(self.leftImageView).offset(LAZSAO750(8.0));
        make.bottom.equalTo(self.leftImageView).offset(-LAZSAO750((8.0)));
        make.centerY.equalTo(self.leftImageView);
    }];
}

- (UIBezierPath *)createPath {
    UIBezierPath *path = [UIBezierPath new];
    
    CGFloat cornerRadius = self.viewModel.model.cornerRadius;
    CGFloat triangleHeight = self.viewModel.model.triangleHeight;
    CGFloat triangleWidth = self.viewModel.model.triangleWidth;
    CGFloat trianglePosition = self.viewModel.model.trianglePosition;
    
    [path moveToPoint:CGPointMake(cornerRadius, 0)];
    
    [path addLineToPoint:CGPointMake(self.bounds.size.width - cornerRadius, 0)];
    // 右上角
    [path addArcWithCenter:CGPointMake(self.bounds.size.width - cornerRadius, cornerRadius)
                        radius:cornerRadius
                    startAngle:M_PI * 3 / 2
                      endAngle:0
                     clockwise:YES];
        
    // 右边
    [path addLineToPoint:CGPointMake(self.bounds.size.width, self.bounds.size.height - cornerRadius - triangleHeight)];

    // 右下角
    [path addArcWithCenter:CGPointMake(self.bounds.size.width - cornerRadius, self.bounds.size.height - cornerRadius - triangleHeight)
                        radius:cornerRadius
                    startAngle:0
                      endAngle:M_PI / 2
                     clockwise:YES];
    // 底边
    [path addLineToPoint:CGPointMake(trianglePosition + triangleWidth, self.bounds.size.height - triangleHeight)];
    // 三角
    [path addLineToPoint:CGPointMake(trianglePosition + triangleWidth / 2, self.bounds.size.height)];
    [path addLineToPoint:CGPointMake(trianglePosition, self.bounds.size.height - triangleHeight)];
    [path addLineToPoint:CGPointMake(cornerRadius, self.bounds.size.height - triangleHeight)];
    // 左下角
    [path addArcWithCenter:CGPointMake(cornerRadius, self.bounds.size.height - cornerRadius - triangleHeight)
                        radius:cornerRadius
                    startAngle:M_PI / 2
                      endAngle:M_PI
                     clockwise:YES];
    // 左边
    [path addLineToPoint:CGPointMake(0, cornerRadius)];
    // 左上角
    [path addArcWithCenter:CGPointMake(cornerRadius, cornerRadius)
                        radius:cornerRadius
                    startAngle:M_PI
                      endAngle:M_PI * 3 / 2
                     clockwise:YES];
    // 关闭路径
    [path closePath];
    return path;
}
@end

File: BubbleView/BubbleView/CustomBubbleView.swift
//
//  CustomBubbleView.swift
//  BubbleView
//
//  Created by Theo on 2024/8/27.
//

import Foundation
import SnapKit


let LAZSAO750 = { (value: CGFloat) -> CGFloat in
    return value * (UIScreen.main.bounds.size.width / DESIGN_WIDTH)
}

class CustomBubbleView: UIView {
    var viewModel: BubbleViewModel?
        
    private lazy var label: UILabel = {
        let label = UILabel()
        return label
    }()
        
    private lazy var leftImageView: UIImageView = {
        let imageView = UIImageView()
        return imageView
    }()
    
    private lazy var shapeLayer: CAShapeLayer = {
        let layer = CAShapeLayer()
        layer.fillColor = UIColor.white.cgColor
        layer.strokeColor = UIColor.white.cgColor
        return layer
    }()
        
    private lazy var gradientContentLayer: CAGradientLayer = {
        let layer = CAGradientLayer()
        layer.mask = shapeLayer
        return layer
    }()
        
    private lazy var borderLayer: CAShapeLayer = {
        let layer = CAShapeLayer()
        return layer
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }
        
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
        
    private func setupView() {
        addSubview(label)
        addSubview(leftImageView)
        
        layer.insertSublayer(gradientContentLayer, at: 0)
        layer.addSublayer(borderLayer)
    }
            
    func bindViewModel(_ viewModel: BubbleViewModel) {
        self.viewModel = viewModel
        
        let attributedString = NSMutableAttributedString()
        if let titleExtend = self.viewModel?.model.titleExtend as? [[String: Any]] {
            for dict in titleExtend {
                if let text = dict["text"] as? String,
                    let colorString = dict["color"] as? String,
                    let textColor = UIColor.lazColor(hex: colorString) {
                    
                    var font = UIFont.systemFont(ofSize: LAZSAO750(24.0))
                    if let isBold = dict["bold"] as? String, isBold == "true" {
                        font = UIFont.boldSystemFont(ofSize: LAZSAO750(22.0))
                    }
                        
                    let attributes: [NSAttributedString.Key: Any] = [
                        .font: font,
                        .foregroundColor: textColor
                    ]
                        
                    let attributedText = NSAttributedString(string: text, attributes: attributes)
                    attributedString.append(attributedText)
                }
            }
        }
        self.label.numberOfLines = 2
        self.label.attributedText = attributedString
            
        self.borderLayer.strokeColor = self.viewModel?.model.borderLineColor?.cgColor
        self.borderLayer.lineWidth = self.viewModel?.model.borderLineSize ?? 2.0
        self.leftImageView.image = UIImage(named: "Image-1614")
            
        self.setupLayer()
        self.setupLayout()
    }
        
    override func layoutSubviews() {
        super.layoutSubviews()
        shapeLayer.path = createPath().cgPath
        gradientContentLayer.frame = bounds
        borderLayer.path = shapeLayer.path
    }
        
    private func setupLayer() {
        gradientContentLayer.colors = viewModel?.model.contentColors
        gradientContentLayer.startPoint = CGPoint(x: 0.0, y: 1.0)
        gradientContentLayer.endPoint = CGPoint(x: 1.0, y: 0.0)
            
        gradientContentLayer.locations = [0.0, 0.4, 0.7, 1.0]
        
        //borderLayer.path = shapeLayer.path
        borderLayer.fillColor = UIColor.clear.cgColor
    }
        
    private func setupLayout() {
        guard let viewModel = viewModel else {return }
        leftImageView.snp.remakeConstraints { make in
            make.height.equalTo(self.snp.height).offset(-LAZSAO750(12.0 + 12.0) - (viewModel.model.triangleHeight))
            make.width.equalTo(leftImageView.snp.height)
            make.left.equalTo(self).offset(LAZSAO750(12.0))
            make.top.equalTo(self).offset(LAZSAO750(12.0))
        }
            
        label.snp.remakeConstraints { make in
            make.right.equalTo(self).offset(-LAZSAO750(36.0))
            make.left.equalTo(leftImageView.snp.right).offset(LAZSAO750(12.0))
            make.top.equalTo(leftImageView).offset(LAZSAO750(8.0))
            make.bottom.equalTo(leftImageView).offset(-LAZSAO750(8.0))
        }
    }

    private func createPath() -> UIBezierPath {
        let path = UIBezierPath()
        guard let viewModel = viewModel else { return path }
            
        let cornerRadius = viewModel.model.cornerRadius
        let triangleHeight = viewModel.model.triangleHeight
        let triangleWidth = viewModel.model.triangleWidth
        let trianglePosition = viewModel.model.trianglePosition
            
        path.move(to: CGPoint(x: cornerRadius, y: 0))
        path.addLine(to: CGPoint(x: bounds.size.width - cornerRadius, y: 0))
            
        // Right Top Corner
        path.addArc(withCenter: CGPoint(x: bounds.size.width - cornerRadius, y: cornerRadius),
                    radius: cornerRadius,
                    startAngle: .pi * 3 / 2,
                    endAngle: 0,
                    clockwise: true)
            
        // Right Side
        path.addLine(to: CGPoint(x: bounds.size.width, y: bounds.size.height - cornerRadius - triangleHeight))
            
        // Right Bottom Corner
        path.addArc(withCenter: CGPoint(x: bounds.size.width - cornerRadius, y: bounds.size.height - cornerRadius - triangleHeight),
                    radius: cornerRadius,
                    startAngle: 0,
                    endAngle: .pi / 2,
                    clockwise: true)
            
        // Bottom Side
        path.addLine(to: CGPoint(x: trianglePosition + triangleWidth, y: bounds.size.height - triangleHeight))
            
        // Triangle
        path.addLine(to: CGPoint(x: trianglePosition + triangleWidth / 2, y: bounds.size.height))
        path.addLine(to: CGPoint(x: trianglePosition, y: bounds.size.height - triangleHeight))
        path.addLine(to: CGPoint(x: cornerRadius, y: bounds.size.height - triangleHeight))
            
        // Left Bottom Corner
        path.addArc(withCenter: CGPoint(x: cornerRadius, y: bounds.size.height - cornerRadius - triangleHeight),
                    radius: cornerRadius,
                    startAngle: .pi / 2,
                    endAngle: .pi,
                    clockwise: true)
            
            // Left Side
        path.addLine(to: CGPoint(x: 0, y: cornerRadius))
            
            // Left Top Corner
        path.addArc(withCenter: CGPoint(x: cornerRadius, y: cornerRadius),
                    radius: cornerRadius,
                    startAngle: .pi,
                    endAngle: .pi * 3 / 2,
                    clockwise: true)
            
        path.close()
        return path
    }
}
